File: 2-AddTwoNumbers.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    def show(self):
        node = self
        while node:
            print(node.val)
            node = node.next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        result = ListNode(0)
        currentNode = result

        carry = 0
        while l1 or l2:
            if l1:
                x = l1.val
                l1 = l1.next
            else:
                x = 0
            if l2:
                y = l2.val
                l2 = l2.next
            else:
                y = 0
            sum = x + y + carry
            carry = sum // 10
            currentNode.val = sum % 10
            if l1 or l2 or carry:
                currentNode.next = ListNode(carry)
                currentNode = currentNode.next
            
        return result


if __name__ == "__main__":

    l1 = ListNode(2)
    l1.next = ListNode(4)
    l1.next.next = ListNode(3)
    # l1.show()

    l2 = ListNode(5)
    l2.next = ListNode(6)
    l2.next.next = ListNode(4)
    # l2.show()

    result = Solution().addTwoNumbers(l1, l2)    
    result.show()

File: C/2.AddTwoNumbers.c
#include <stdio.h>
#include <stdlib.h>

/**
 * Definition for singly-linked list.
 */
struct ListNode
{
    int val;
    struct ListNode *next;
};
typedef struct ListNode ListNode;

void showListNode(ListNode list)
{

    while (1) {
        printf("%d ", list.val);
        if (list.next == NULL) break;
        list = *list.next;
    }
    printf("\n");
}

struct ListNode *addTwoNumbers(struct ListNode *l1, struct ListNode *l2) 
{
    int x, y, sum, carry = 0;
    struct ListNode *resultHead = NULL, *resultTail = NULL;

    while (1) {
        x = 0;
        y = 0;

        if (l1 == NULL && l2 == NULL) break;
        if (l1 != NULL) x = l1->val;
        if (l2 != NULL) y = l2->val;
        sum = x + y + carry;

        struct ListNode *newNode = (struct ListNode *)malloc(sizeof(struct ListNode));
        newNode->val = sum % 10;
        newNode->next = NULL;
        carry = sum / 10;

        if (resultHead == NULL) {
            resultHead = newNode;
            resultTail = resultHead;
        }
        else {
            resultTail->next = newNode;
            resultTail = resultTail->next;
        }

        if (l1 != NULL) l1 = l1->next;
        if (l2 != NULL) l2 = l2->next;
    }

    if (carry) {
        struct ListNode *newNode = (struct ListNode *)malloc(sizeof(struct ListNode));
        newNode->val = carry;
        newNode->next = NULL;
        resultTail->next = newNode;
        resultTail = resultTail->next;
    }

    return resultHead;
}

int main()
{
    struct ListNode l1_1 = {9, NULL};
    struct ListNode l1_2 = {4, NULL};
    struct ListNode l1_3 = { 3, NULL};
    l1_1.next = &l1_2;
    l1_2.next = &l1_3;
    // showListNode(l1_1);

    struct ListNode l2_1 = {1, NULL};
    struct ListNode l2_2 = {6, NULL};
    struct ListNode l2_3 = {4, NULL};
    l2_1.next = &l2_2;
    l2_2.next = &l2_3;
    // showListNode(l2_1);

    struct ListNode *result = addTwoNumbers(&l1_1, &l2_1);
    showListNode(*result);
    // TODO: free ListNode;
}